### DISCUSSION TITLE
IntelliSense: Avoid reparsing active document possible?

### DISCUSSION THREAD

**Discussion by juxeii** [microsoft/vscode-cpptools#8881] [Q&A]
We have a very large code base.
One problem I have with Intellisense is that it reparses the active document although it has been parsed before.
Here is an example of how I cannot work in a smooth fashion:
I mark one include file A and click "Go to Definition".
The file A opens in an instant which is fine.
Now I go back to the original file from where I jumped from and browse to other include files.
Now I want to browse to file A again(no changes were made). It can take many seconds now to open the file A again.

I wonder why is that? Why is there no cache used which opens the file A again in an instant?
On the console for C/C++ I can see "idle loop: reparsing the active document".
Is there something I can do to improve the user experience here?

**Comment by sean-mcmanus** [comment]
There are multiple different types of parsing occurring. I'm not sure which is causing a problem for you.

What do you mean by "it can take many seconds now to open file A" -- how are you opening file A? Our parsing should not block file opening.

Are you saying IntelliSense is taking a while to update? Do you see message Update IntelliSense time (sec): (when C_Cpp.loggingLevel is set to "Debug")? How large is the seconds? There is some caching going on (the IntelliSense process still exists), but I'm looking into if that reparse is always required or not when the active document is changed and no edit has occurred...

**Comment by juxeii** [comment]
Sorry to be not very precise. By opening file A I mean, that I again mark the include directive with file A, and VSC "jumps" to file A when I hit "Go To Definition".
I see the message "Update IntelliSense" and for me it takes sometimes ~20sec for the file A to open after hitting "Go To Definition". The problem I have is, that I had jumped(Go To Definition) to file A before in my session. There was an entry with "Update IntelliSense" which took 0.1 seconds the first time I jumped to A. Why does the "Go To Definiton" then take 20 seconds for a later "rejump" to file A?

It would just seem logical to me that when you have executed a "Go To Definiton" for an include directive, that all later executions will not take far longer than the first one. Especially not when you just browse the code and make not any changes.

Edit: This high increase in "Go To Definition" time occurs only sometimes and occurs more likely the more you have "jumped" to other include files in between.

**Reply by sean-mcmanus** [reply]
I repro the issue. It's possible we may be able to skip the parse update, which is primarily there in case a dependent header was changed externally or some other bug was hit. I'll move this to an Issue. Investigating...

**Comment by juxeii** [comment]
Thank you very much. In fact this is the only reason I cannot use VSC on big projects and I still rely on Clion for now. If you could improve just a little here that would be awesome :)

**Reply by sean-mcmanus** [reply]
This issue discussion got moved to https://github.com/microsoft/vscode-cpptools/issues/8885 .
